### PR TITLE
Add the laws implementations

### PR DESCRIPTION
**Pull request by SimonRichardson**
The motivation behind this is to allow any person wanting to implement
any structure that should conform to the specification will also have
the readme and more importantly the implementations of the laws.

The advantage of this as well is that any changes to the spec should be
reflected in the laws folder.

### CODE REVIEW

**Comment by SimonRichardson** [comment]
Note: not all laws implemented yet, more to come, but wondered what people thought about this PR?

File: laws/applicative.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+const {identity, apply} = require('fantasy-combinators');
+const {ap} = require('../index');
```
Comment (davidchambers) [inline laws/applicative.js]: `require('..')` is sufficient here. :)

File: laws/applicative.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+const id = (t) => (eq) => (x) => {
```
Comment (davidchambers) [inline laws/applicative.js]: These parens are all optional.

File: laws/applicative.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    return eq(a, b);
+};
+
+const interchange = (t) => (eq) => (x) => {
+    const a = t.of(identity).ap(t.of(x));
+    const b = t.of(apply(t)).ap(t.of(x));
+    return eq(a, b);
+};
+
+modules.exports = { identity: id, homomorphism, interchange };
```
Comment (davidchambers) [inline laws/applicative.js]: [`\n`](https://github.com/editorconfig/editorconfig/wiki/Newline-at-End-of-File-Support)?

**Comment by davidchambers** [comment]
Including these files is a great idea!

File: laws/applicative.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+const {identity, apply} = require('fantasy-combinators');
+const {ap} = require('../index');
+
+const id = (t) => (eq) => (x) => {
+    const a = t.of(identity).ap(t.of(x));
```
Comment (rpominov) [inline laws/applicative.js]: This should be `t[of](identity)[ap](t[of](x))` I think. 

Btw we have a problem with `of` here, it's a reserved word, we can't use it for variable name :crying_cat_face:

Comment (SimonRichardson) [inline laws/applicative.js]: Didn't know that! OK will change

File: laws/applicative.js (hunk 1 above)
Comment (SimonRichardson) [inline laws/applicative.js]: Learned something new today :dancer:

File: laws/applicative.js (hunk 2 above)
Comment (SimonRichardson) [inline laws/applicative.js]: Yeah, I'm still under sided if they should be required for consistency. I'll remove for now.

File: laws/applicative.js (hunk 3 above)
Comment (SimonRichardson) [inline laws/applicative.js]: not fussed either way.

**Comment by joneshf** [comment]
This is a great idea for most of the laws, but some might require more thought. For instance, the composition law for `Functor`. It will not suss out a bad `Functor` all the time. I know, these are tests, not proofs so you can't find all the issues, but take the typical "maybe" implementation that springs up fairly often:

```js
function NullHelper(x) {
  return {
    map: f => x == null ? NullHelper(x) : NullHelper(f(x)),
  };
}
```

While it passes both the tests given, that's definitely not a `Functor`.

**Comment by rpominov** [comment]
Hm, why we allow specify arbitrary `x`, btw? Can we check laws with our own values, like `0, 1, null, undefined` (or event do quick-check)? Is there cases when a type can work only with limited set of values?

**Comment by SimonRichardson** [comment]
@rpominov The idea is I want to get quick check to call these, so we need to pass in values.

**Comment by rpominov** [comment]
Quick check should catch @joneshf's example then, I think.

**Comment by SimonRichardson** [comment]
fyi: i've moved this discussion to the gitter room, if anybody wants to join me?

**Comment by SimonRichardson** [comment]
@DrBoolean I'm having a problem getting the `Id` implementation of `sequence` to validate without changing the identity law.

The specification requires this:
```
2. `u.map(x => Id(x)).sequence(Id.of)` is equivalent to `Id.of` (identity)
```
Is it not?
```
2. `u.map(identity).sequence(Id.of)` is equivalent to `Id.of` (identity)
```

Or am I missing what `u` should be?
https://github.com/fantasyland/fantasy-land/blob/0540c4206665ea3f041a556ce1fa461bdc3fdde2/id_test.js#L98

**Comment by SimonRichardson** [comment]
I've also implemented `id_test` so that we can verify that the laws pass with `Id`. We can also pass this through travis to make sure that nothing broke between spec updates.

**Comment by rpominov** [comment]
I think this law should be:

```
2. `u.map(x => Id(x)).sequence(Id.of)` is equivalent to `Id.of(u)` (identity)
```

So, for example, `u = [1,2,3]`, we can either do this:

```js
u.map(Id) // [Id(1),Id(2),Id(3)]
  .sequence(Id.of) // Id([1,2,3])
```

Or just:

```js
Id.of(u) // Id([1,2,3])
```

**Comment by SimonRichardson** [comment]
Should `Id` be able to sequence over itself?

**Comment by SimonRichardson** [comment]
@rpominov so I got it to pass, it's just a bit weird that `Id` can't sequence over itself.

**Comment by DrBoolean** [comment]
Hi!

@rpominov is right. `u` is another functor and I accidentally wrote the `Id.of` pointfree which is confusing and not even equivalent.

I think the confusion is that in `sequence(Id.of)`, the `Id.of` part is ignored

**Comment by DrBoolean** [comment]
The reason for the argument is for failure. So we're not sequencing Id over itself, but rather supplying type help.

`Some(x).sequence(Id.of) == Id(Some(x))`
vs
`None.sequence(Id.of) == Id(None)`

**Comment by SimonRichardson** [comment]
@DrBoolean makes sense once I've implemented it :dancer: Everybody happy for me to merge?

**Comment by DrBoolean** [comment]
Heck yeah! Want me to add that `Id.of(u)` to the spec real quick?

**Comment by SimonRichardson** [comment]
Yes please. Just add it to this branch.

**Pull request merged**